File: R/superdelta.R
is.wholenumber<-function(x, tol = .Machine$double.eps^0.5){
  res<-numeric(length(x))
  for (i in 1:length(x)){
    res[i]<-ifelse(!is.numeric(x[i]), 0, abs(x[i]-round(x[i]))<tol)
  }
  return(as.logical(res))
}
### turn a vector of t-statistics into two sided pvalues
#Define a function to handle 2-sided and/or 1-sided test#
.t2p <- function(x, df, side=c("two.sided", "less", "greater")){
    if (side=="two.sided"){
        return(2-2*pt(abs(x),df))
    } else if (side=="less"){
        return(pt(x, df))
    } else if (side=="greater"){
        return(1-pt(x, df))
    } else {
        stop(paste("side must be one of the following three options: two.sided (default), less, or greater."))
    }
}

## Median; fold; trim; median
#Trim the outliers on both sides by MAD (Median Absolute Distance)#
#YL made this note on Jul 13, 2015#
.mftm <- function(tvec, trim=0.2){
    tvec.no.na <- na.omit(tvec)
    ## tvec.folded is firsrt centered around median and then folded
    ## around zero.
    tvec.folded <- abs(tvec.no.na-median(tvec.no.na))
    ## Remove trim amount of extreme observations from tvec.folded
    t.est <- median(tvec.no.na[tvec.folded<=quantile(tvec.folded, 1-trim)])
    ##t.est is the median after trimming
    return(t.est)
}

## cluster tvec by mclust
.tclust1 <- function(tvec, ...) {
    ## get rid of the NAs first
    tvec <- na.omit(tvec)
    mm.i <- summary(mclustBIC(tvec, G=3, modelName="V", ...), tvec)
    NC1 <- sum(mm.i$classification==1)
    NC2 <- sum(mm.i$classification==2)
    NC3 <- sum(mm.i$classification==3)
    largest.cluster <- which.max(c(NC1, NC2, NC3))
    t.est <- as.double(mm.i$parameters$mean[largest.cluster])
    return(t.est)
}

.tclust2 <- function(tvec, trim=0.2, ...){
    ## mclust is not that robust, so use the following wrapper as a backup.
    .substitute <- function(x) {
        warning("Mclust does not converge, use the robust method instead.")
        .mftm(tvec, trim=trim)
    }
    t.est <- tryCatch(.tclust1(tvec, ...), warning=.substitute, error=.substitute)
    #If no condition is signaled when evaluating expr#
    #then tryCatch returns the value of the expression#
    return(t.est)
}

## this function takes tmat, which is a matrix of
## t-statistics. tmat[i,j] is the tstat associated with the ith gene
## normalized by the jth gene.  It estimates one best t-stat according
## to different methods.
.est.t <- function(tmat, method=c("robust", "mean", "median", "mclust"), trim=0.2, ...){
    method=match.arg(method)
    if (method=="robust"){
        t.est <- apply(tmat, 1, .mftm, trim=trim)
    } else if (method=="median"){
        t.est <- apply(tmat, 1, median, na.rm=TRUE)
    } else if (method=="mean"){
        t.est <- rowMeans(tmat, na.rm=TRUE)
    } else if (method=="mclust"){
        t.est <- apply(tmat, 1, .tclust2)
    } else {
        stop(paste("Method", method, "is not implemented!"))
    }
    ## 7/30/2011.  Try t-stat with discounted variance. This discount
    ## is useful because the variance of delta is twice as much as the
    ## variance of the original expression.
    return(sqrt(2)*t.est)
}

superdelta <- function(X, classlabel, test="t", side=c("two.sided", "less", "greater"), methods="robust", trim=0.2, baseline="auto", ...)
{
    test <- match.arg(test)
    side <- match.arg(side)
    ngenes <- dim(X)[1]  #nrow#
    nslides <- dim(X)[2]  #ncol#
    teststat <- matrix(0, nrow=ngenes, ncol=length(methods))
    ind <- matrix(0, nrow=ngenes, ncol=length(methods))
    colnames(teststat) <- colnames(ind) <- methods
    rownames(teststat) <- rownames(ind) <- rownames(X)
    ## Let's turn the classlabel into a binary vector that is required
    ## by two_sample_tstat() in C.
    classlabel <- as.integer(as.factor(classlabel)) - 1
    Xa <- X[, classlabel==0]; Xb <- X[, classlabel==1]
    n1 <- ncol(Xa); n2 <- ncol(Xb)
    ## gene-order is calculated via pooled sample variance/STD
    Va <- apply(Xa, 1, var); Vb <- apply(Xb, 1, var)
    DFnum <- (Va/n1 + Vb/n2)^2
    DFdenom <- Va^2/n1^2/(n1-1) + Vb^2/n2^2/(n2-1)
    DF <- round(DFnum/DFdenom)
    ## Now deal with the baseline genes.  First, if baseline is a
    ## vector of genes/names, we assume it is a pre-specified list of
    ## baseline genes.
    if (length(baseline)>1){
        ##assume this is a vector of pre-selected baseline genes (heuristic)
        if (!all(is.wholenumber(baseline))){
            ## try to translate the list of gene symbols into indices
            baseline.genes <- which(rownames(X) %in% baseline)
        } else {
            ## assume that this is a list of integer-valued indices
            baseline.genes <- baseline
        }
    } else if (baseline=="auto") {
        ##if ngenes is greater than 1000, we only use 1000
        ##genes. Otherwise, we use all genes.
        if (ngenes>1000){
            baseline.genes <- sample(ngenes, 1000)
        } else {
            ##use all genes
            baseline.genes <- 1:ngenes
        }
    } else if (baseline=="all"){
        baseline.genes <- 1:ngenes
    } else {
        stop("Variable baseline must be either: 1. A vector of pre-specified baseline genes (length>1). 2. String value 'auto' or 'all'. ")
    }
    ## We have to reshape the data matrix in order to use C function
    ## superdelta_tstats().
    glist <- c(baseline.genes, setdiff(1:ngenes, baseline.genes))
    ## Y must be passed to C program as a numerical VECTOR!
    Y <- as.double(as.matrix(X)[glist, ])
    ## The real calculation starts
    Tmat <- rep(0, ngenes*length(baseline.genes))
    res<-.C("superdelta_tstats",as.double(Y), as.integer(ngenes),
            as.integer(nslides),as.integer(length(baseline.genes)),
            as.integer(classlabel),
            Tmat=as.double(Tmat), PACKAGE="superdelta")$Tmat
    ## Now make a matrix out of the res and re-organize it in the
    ## original ordering.
    tmat <- matrix(res, nrow=ngenes)[order(glist),]
    
    ## Now estimate the centers
    for (mm in methods){
        teststat[,mm] <- .est.t(tmat, method=mm, trim=trim, ...)
    }
    ## Now find the indices of matching pair gene of each gene
    for (mm in methods){
      for (i in 1:ngenes){
        ind[i,mm] <- baseline.genes[which.min(abs(sqrt(2)*tmat[i,]-teststat[i,mm]))]
      }
    }
    ## two sided or one sided test; different test statistics, etc ##
    if (test=="t"){
        rawp=.t2p(teststat, df=DF, side=side)
    } else {
        stop(paste("Test statistic", test, "is not implemented yet!"))
    }

    ## two representations ##
    if (ncol(teststat)==1){
        rr <- data.frame(teststat=teststat, rawp=rawp, ind=ind)
        #rr <- list(teststat=teststat, rawp=rawp, ind=ind, tmat=tmat)
        colnames(rr) <- c("teststat", "rawp", "pairing gene index")
        return(rr)
    } else {
        return(list(teststat=teststat, rawp=rawp, ind=ind))
    }
}

File: src/superdelta.c
#include <R.h>                  /* for NA_REAL and Rmath.h */

/* This is the tolerance level */
const double EPSILON=0.0000001;

/* This function computes two-sample t-stat (without Welch correction) */
double two_sample_tstat(double *Y, const int *CL, const int *Ylen)
{
  double mean_no_na[2]={0,0},ss_no_na[2]={0,0},devi;
  double c0;
  int i,count[2]={0,0},class;
  double num, denom;   /* numerator and denominator */

  /*compute the mean and count first*/
  /* count is the number of objects in each class*/
  for (i=0; i<*Ylen; i++) {
    if (Y[i]==NA_REAL)          /* simply ignore NAs */
      continue;
    class=CL[i];
    mean_no_na[class] += Y[i];
    count[class]++;
  }
  mean_no_na[0]/=(count[0]*1.0);
  mean_no_na[1]/=(count[1]*1.0);

  /*compute the variance in each group*/
  for (i=0; i<*Ylen; i++) {
    if (Y[i]==NA_REAL)
      continue;
    class=CL[i];
    devi=(Y[i]-mean_no_na[class]);
    ss_no_na[class] += devi*devi;
  }

  /* This is the equal-var t-stat; not the Welch-corrected stat. Note that the sign is compatible with genefilter/t.test(), not multtest. */
  /* num=mean_no_na[1]-mean_no_na[0]; */
  num=mean_no_na[0]-mean_no_na[1];
  c0= (double)(count[0]+count[1])/(count[0]*count[1]*(count[0]+count[1]-2));
  denom=sqrt((ss_no_na[0]+ss_no_na[1])*c0);
  /* check the denom. */
  if(denom<EPSILON)
    return NA_REAL; /*If too small denominator, treat as NA*/
  return num/denom;
}

/* This function computes two-sample t-stat (with Welch correction) */
double two_sample_tstatWel(double *Y, const int *CL, const int *Ylen)
{
  double mean_no_na[2]={0,0},ss_no_na[2]={0,0},devi;
  /*double df;*/
  int i,count[2]={0,0},class;
  double num, denom;   /* numerator and denominator */

  /*compute the mean and count first*/
  /* count is the number of objects in each class*/
  for (i=0; i<*Ylen; i++) {
    if (Y[i]==NA_REAL)          /* simply ignore NAs */
      continue;
    class=CL[i];
    mean_no_na[class] += Y[i];
    count[class]++;
  }
  mean_no_na[0]/=(count[0]*1.0);
  mean_no_na[1]/=(count[1]*1.0);

  /*compute the variance in each group*/
  for (i=0; i<*Ylen; i++) {
    if (Y[i]==NA_REAL)
      continue;
    class=CL[i];
    devi=(Y[i]-mean_no_na[class]);
    ss_no_na[class] += devi*devi;
  }

  /* This is the Welch-corrected t-stat. Note that the sign is compatible with genefilter/t.test(), not multtest. */
  /* num=mean_no_na[1]-mean_no_na[0]; */
  num=mean_no_na[0]-mean_no_na[1];
  /* c0= (double)(count[0]+count[1])/(count[0]*count[1]*(count[0]+count[1]-2)); */
  denom=sqrt(ss_no_na[0]/(count[0]-1.0)/count[0]+ss_no_na[1]/(count[1]-1.0)/count[1]);
  /* check the denom. */
  if(denom<EPSILON)
    return NA_REAL; /*If too small denominator, treat as NA*/
  return num/denom;
}

/* compute delta (difference) from two genes Y1, Y2 and then
   return two-sample t-statistic. */
double delta_tstat(const double *Y1, const double *Y2, const int *nslides, const int *CL)
{
  /* assign a vector of size *nslides to hold delta */
  double delta[*nslides];
  double t;
  int i;
  for(i=0; i<*nslides; i++){
    delta[i]=Y1[i]-Y2[i];
  }
  /* Now compute the t-stats */
  t=two_sample_tstatWel(delta, CL, nslides);
  return t;
}

/* Function superdelta_stats_ttest() below takes gene expression data (Y), a vector of indices of baseline genes (basegenes),
 a vector of group class labels (CL), and a pre-defined matrix Tmat. Note that Y must be organized in such a way that:
 1. The first nbase genes are the baseline genes;
 2. CL must be a 0,1 vector;
 3. Tmat must be a ngenes * nbase matrix. */
void superdelta_tstats(const double *Y, const int *ngenes, const int *nslides, const int *nbase, const int *CL, double *Tmat)
{
  int n=*nslides, p=*ngenes;
  int i,j;
  /* First, let's copy the vectorized Y into 2d-array in C (Yc). */
  double **Yc;
  Yc = (double **) R_alloc(p, sizeof(double));
  for(i = 0; i < p; i++) {
    Yc[i] = (double *) R_alloc(n, sizeof(double));
    for(j = 0; j < n; j++){
      Yc[i][j] = Y[j*p+i];
    }
  }

  /* Now let's compute the (nbase-1) x (nbase-1) dim submatrix of
     Tmat, which is symmetric. */
  for(i=0; i<*nbase; i++){
    /* I decide to assign NA to diagonal elements. */
    Tmat[i*p+i]=NA_REAL;
    /* Off-diagonal elements; use symmetry to save time */
    for(j=0;j<i;j++){
      Tmat[j*p+i]=delta_tstat(Yc[i], Yc[j], nslides, CL);
      Tmat[i*p+j]= -Tmat[j*p+i];
    }
  }
  /* The rest of the Tmat; no symmetry. */
  for(i=*nbase; i<p; i++){
    for(j=0;j<*nbase;j++){
      Tmat[j*p+i]= delta_tstat(Yc[i], Yc[j], nslides, CL);
    }
  }
}

/* A convenient function that can be used in place of rowttest() in genefilter package */
void fastt(const double *Y, const int *ngenes, const int *nslides, const int *CL, double *Tvec)
{
  int n=*nslides, p=*ngenes;
  int i,j;
  /* First, let's copy the vectorized Y into 2d-array in C (Yc). */
  double **Yc;
  Yc = (double **) R_alloc(p, sizeof(double));
  for(i = 0; i < p; i++) {
    Yc[i] = (double *) R_alloc(n, sizeof(double));
    for(j = 0; j < n; j++){
      Yc[i][j] = Y[j*p+i];
    }
  }

  /* Now let's compute the two-sample t-statistics (with Welch adjustment). */
  for(i=0; i<*ngenes; i++){
    Tvec[i]=two_sample_tstatWel(Yc[i], CL, nslides);
  }
}
